### PR TITLE
BUGFIX: Don't re-publish already existing resources

### PR DESCRIPTION
**Pull request by dimaip**
Uses the v2 ListObjects API: https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-s3-2006-03-01.html#listobjectsv2

### CODE REVIEW

**Comment by dimaip** [comment]
Would be nice to optimize object traversal in this style: https://github.com/neos/flow-development-collection/commit/879fba19f93d0a8628682698e57da9f1b58ad7d4
But could be done in a different PR.

File: Classes/S3Target.php [hunk 1]
```diff
@@ -191,31 +192,37 @@ public function publishCollection(CollectionInterface $collection)
...
-                    'ContentType' => $object->getMediaType(),
-                    'MetadataDirective' => 'REPLACE',
-                    'Key' => $objectName
-                );
-                try {
-                    $this->s3Client->copyObject($options);
-                } catch (S3Exception $e) {
-                    throw new Exception(sprintf('Could not copy resource with SHA1 hash %s of collection %s from bucket %s to %s: %s', $object->getSha1(), $collection->getName(), $storageBucketName, $this->bucketName, $e->getMessage()), 1431009234);
+                if (array_key_exists($objectName, $obsoleteObjects)) {
+                    $this->systemLogger->log(sprintf('The resource object "%s" (MD5: %s) has already been published to bucket "%s", no need to re-publish', $objectName, $this->bucketName), LOG_DEBUG);
```
Comment (kdambekalns) [inline Classes/S3Target.php]: The MD5 parameter is not given.

File: Classes/S3Target.php [hunk 2]
```diff
@@ -167,20 +167,21 @@ public function publishCollection(CollectionInterface $collection)
...
+                }
                 if ($result->get('IsTruncated')) {
-                    $requestArguments['marker'] = $result->get('NextMarker');
+                    $requestArguments['ContinuationToken'] = $result->get('NextContinuationToken');
                 }
             } while ($result->get('IsTruncated'));
         }
 
-        $obsoleteObjects = array_fill_keys(array_keys($this->existingObjectsInfo), true);
+        $obsoleteObjects = array_fill_keys($this->existingObjectsInfo, true);
```
Comment (kdambekalns) [inline Classes/S3Target.php]: Now that this is used to skip existing objects, the name `$obsoleteObjects` is very misleading. Rather use something like `$objectsAlreadyInBucket`…

Comment (kdambekalns) [reply Classes/S3Target.php]: Yes, but it's not _only_ obsolete objects. I just have this scenatio in mind: sometime, someone tries to optimize things and moves the removal of "obsolete" objects around so it happens before the `unset()` call(s)… and 💥 there go your files.

Also, naming for the sake of itself. Honestly, this is important. Maybe `$potentiallyObsoleteObjects`?

File: Classes/S3Target.php (hunk 1 above)
Comment (dimaip) [reply Classes/S3Target.php]: Oups! Fixed.

File: Classes/S3Target.php (hunk 2 above)
Comment (dimaip) [reply Classes/S3Target.php]: Will fix

**Comment by dimaip** [comment]
@kdambekalns I've fixed your comments, also fixed an upmerge error

**Comment by kdambekalns** [comment]
Thanks for fixing. I merged in GitHub, had to do it twice, only to get a second empty commit. Sigh. Local clone and a proper IDE FTW. :)

**Comment by dimaip** [comment]
My bad, did this PR in the middle of a lot of stuff and just copied by hand to Github... Will try to deliver more quality next time ;)

**Pull request merged**